### ISSUE TITLE
Windows 11 - CUDA + onnxruntime setup

### ISSUE THREAD

**Issue by mosure** [pykeio/ort#171]
I am having some difficulty with adding CUDA/openvino execution providers.

```
[dependencies.ort]
version = "2.0.0-alpha.4"
# I've also tried `git = "https://github.com/pykeio/ort.git"`, `2.0.0-rc.0`
optional = true
default-features = false
features = [
  "cuda",
  "load-dynamic",
  "ndarray",
  "openvino",
]
```

I've tried both [onnxruntime-win-x64-cuda12-1.17.1.zip](https://github.com/microsoft/onnxruntime/releases/download/v1.17.1/onnxruntime-win-x64-cuda12-1.17.1.zip) and [onnxruntime-win-x64-gpu-1.17.1.zip](https://github.com/microsoft/onnxruntime/releases/download/v1.17.1/onnxruntime-win-x64-gpu-1.17.1.zip) pre-built releases, using the following environment:
```
ONNXRUNTIME_DIR="C:\Users\hctim\repos\bevy_ort\onnxruntime\" ORT_DYLIB_PATH="C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_openvino.dll" cargo run
```

```
❯ echo $CUDA_HOME
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA
12.2
❯ echo $CUDNN_HOME
C:\Program Files\NVIDIA\CUDNN
9.0
```

### execution provider library loading error
```
2024-03-09T18:17:08.521250Z  INFO ort: Loaded ONNX Runtime dylib with version '1.17.1'
2024-03-09T18:17:08.615835Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `CoreMLExecutionProvider`: Error trying to load symbol `OrtSessionOptionsAppendExecutionProvider_CoreML` from dynamic library: GetProcAddress failed
2024-03-09T18:17:08.626525Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `CUDAExecutionProvider`: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1209 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_cuda.dll"

2024-03-09T18:17:08.627192Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `OpenVINOExecutionProvider`: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1209 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_openvino.dll"

2024-03-09T18:17:08.627943Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `DmlExecutionProvider`: Error trying to load symbol `OrtSessionOptionsAppendExecutionProvider_DML` from dynamic library: GetProcAddress failed
2024-03-09T18:17:08.630242Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `TensorrtExecutionProvider`: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1209 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_tensorrt.dll"

2024-03-09T18:17:08.630606Z  INFO apply_execution_providers: ort::execution_providers: Successfully registered `CPUExecutionProvider`
```

**Comment by decahedron1** [comment]
> `ORT_DYLIB_PATH="C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_openvino.dll"`

This is not valid. `ORT_DYLIB_PATH` must point to the `onnxruntime.dll`, not the provider DLLs as those only implement the EPs and not the ORT API itself.

**Comment by mosure** [comment]
I see, thanks for clarifying. there is a similar behavior with `ORT_DYLIB_PATH=.../onnxruntime.dll`:

```
❯ ORT_DYLIB_PATH="C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime.dll" cargo run
...
2024-03-09T19:30:13.299127Z  INFO ort: Loaded ONNX Runtime dylib with version '1.17.1'
2024-03-09T19:30:13.394502Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `CoreMLExecutionProvider`: Error trying to load symbol `OrtSessionOptionsAppendExecutionProvider_CoreML` from dynamic library: GetProcAddress failed
2024-03-09T19:30:13.501907Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `CUDAExecutionProvider`: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1209 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_cuda.dll"

2024-03-09T19:30:13.502453Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `OpenVINOExecutionProvider`: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1209 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Users\hctim\repos\bevy_ort\onnxruntime\onnxruntime_providers_openvino.dll"

2024-03-09T19:30:13.502789Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `DmlExecutionProvider`: Error trying to load symbol `OrtSessionOptionsAppendExecutionProvider_DML` from dynamic library: GetProcAddress failed
2024-03-09T19:30:13.503113Z  WARN apply_execution_providers: ort::execution_providers: An error occurred when attempting to register `TensorrtExecutionProvider`: TensorRT execution provider is not enabled in this build.       
2024-03-09T19:30:13.503340Z  INFO apply_execution_providers: ort::execution_providers: Successfully registered `CPUExecutionProvider`
```

**Comment by decahedron1** [comment]
There are additional prerequisites for OpenVINO, it's possible it can't find a DLL thus the `LoadLibrary` error: https://onnxruntime.ai/docs/execution-providers/OpenVINO-ExecutionProvider.html

Also: is `onnxruntime_providers_shared.dll` present in the `onnxruntime` folder?

**Comment by mosure** [comment]
the `onnxruntime` folder looks like this, copied from here: [onnxruntime-win-x64-cuda12-1.17.1.zip](https://github.com/microsoft/onnxruntime/releases/download/v1.17.1/onnxruntime-win-x64-cuda12-1.17.1.zip)
```
❯ cd onnxruntime
❯ ls
onnxruntime.dll  onnxruntime_providers_cuda.dll  onnxruntime_providers_shared.dll
onnxruntime.lib  onnxruntime_providers_cuda.lib  onnxruntime_providers_shared.lib
onnxruntime.pdb  onnxruntime_providers_cuda.pdb  onnxruntime_providers_shared.pdb
```

**Comment by decahedron1** [comment]
I can't reproduce this (Windows 11 + CUDA 12.1) with the `onnxruntime-win-x64-cuda12-1.17.1.zip` you uploaded:
```pwsh
$env:ORT_DYLIB_PATH = 'E:\Downloads\onnxruntime-win-x64-cuda12-1.17.1\onnxruntime-win-x64-cuda-1.17.1\lib\onnxruntime.dll';
cargo run --features cuda
```

CUDA is registered successfully. Maybe there's an issue with your CUDA install? Is cuDNN in the `PATH`? Maybe `zlibwapi.dll` is missing?

`onnxruntime_providers_openvino.dll` has nothing to do with Microsoft's CUDA binaries since they weren't compiled with OpenVINO support. If you want to have access to both OpenVINO + CUDA you'd have to compile ONNX Runtime from source; simply placing the EP DLL there unfortunately doesn't work.

**Comment by mosure** [comment]
I've resolved this by downgrading cuDNN from v9.0 -> v8.9.7. thanks for the help!

**Comment by deadash** [comment]
Where can I find onnxruntime_providers_openvino.dll
